src/Workspaces/Core/Portable/FlowAnalysis/CustomDataFlowAnalysis.cs

```diff
@@ -45,10 +45,7 @@ public static TBlockAnalysisData Run(ControlFlowGraph controlFlowGraph, DataFlow
             }
 
             var initialAnalysisData = analyzer.GetCurrentAnalysisData(blocks[0]);
-            if (initialAnalysisData == default)
-            {
-                initialAnalysisData = analyzer.GetEmptyAnalysisData();
```

Comment:
@mavasani for review of the change to `CustomDataFlowAnalysis.cs`, `DataFlowAnalyzer.cs` and `BasicBlockReachabilityDataFlowAnalyzer.cs`.
For background, the comparison of `default` to an unconstrained type was improperly allowed (and interpreted as a comparison against `null`). With this PR, this becomes an error.

Reply:
Hum, I just realized my fix is wrong :-( I'll take another look

Reply:
Yes, the intent is to allow the type parameter to be either a struct or class, and if it is default value then initialize it by invoking the GetEmptyAnalysisData API. What would be the recommended alternative here?

Reply:
@mavasani The way I fixed this is to require `GetCurrentAnalysisData` to return a value (possibly an empty one). 

---
In reply to: [314401113](https://github.com/dotnet/roslyn/pull/37596#discussion_r314401113) [](ancestors = 314401113)

Reply:
I see that you updated the comment on the abstract method `GetCurrentAnalysisData`, but did not change the override which may return null.
https://github.com/dotnet/roslyn/blob/7eb2327b39d0abfd9a4385e6c20777aaf551301f/src/Workspaces/Core/Portable/FlowAnalysis/SymbolUsageAnalysis/SymbolUsageAnalysis.DataFlowAnalyzer.cs#L149-L150

Can you change the override to `_analysisData.GetBlockAnalysisData(basicBlock) ?? GetEmptyAnalysisData();`?